azurelinuxagent/common/protocol/wire.py

```diff
@@ -826,35 +745,39 @@ def save_goal_state(incarnation, xml_text):
 
         raise ProtocolError("Exceeded max retry updating goal state")
 
-    def _set_goal_state(self, new_goal_state):
-        if new_goal_state is None:
-            logger.warn("Setting empty Goal State object!")
-        self._goal_state = new_goal_state
+    def _update_host_plugin(self, container_id, role_config_name):
+        if self._host_plugin is not None:
+            self._host_plugin.update_container_id(container_id)
+            self._host_plugin.update_role_config_name(role_config_name)
 
-    def _set_hosting_env(self, new_hosting_env):
-        if new_hosting_env is None:
-            logger.warn("Setting empty Hosting Environment object!")
-        self._hosting_env = new_hosting_env
+    def _save_goal_state(self):
+        try:
+            self.goal_state_flusher.flush(datetime.utcnow())
 
-    def _set_shared_conf(self, new_shared_config):
-        if new_shared_config is None:
-            logger.warn("Setting empty Shared Config object!")
-        self._shared_conf = new_shared_config
+        except Exception as e:
+            logger.warn("Failed to save the previous goal state to the history folder: {0}", ustr(e))
+
+        try:
+            local_file = os.path.join(conf.get_lib_dir(), INCARNATION_FILE_NAME)
+            self.save_cache(local_file, self._goal_state.incarnation)
 
-    def _set_certs(self, new_certs):
-        if new_certs is None:
-            logger.warn("Setting empty Certs object!")
-        self._certs = new_certs
+            def save_if_not_none(goal_state_property, file_name):
+                file_path = os.path.join(conf.get_lib_dir(), file_name)
 
-    def _set_ext_conf(self, new_ext_conf):
-        if new_ext_conf is None:
-            logger.warn("Setting empty Extension Config object!")
-        self._ext_conf = new_ext_conf
+                if goal_state_property is not None and goal_state_property.xml_text is not None:
+                    self.save_cache(file_path, goal_state_property.xml_text)
+                else:  # remove the file from the previous goal state if it exists
+                    if os.path.exists(file_path):
+                        os.remove(file_path)
 
-    def _set_remote_access(self, new_remote_access):
-        if new_remote_access is None:
-            logger.warn("Setting empty Remote Access object!")
-        self._remote_access = new_remote_access
+            save_if_not_none(self._goal_state, GOAL_STATE_FILE_NAME.format(self._goal_state.incarnation))
+            save_if_not_none(self._goal_state.hosting_env, HOSTING_ENV_FILE_NAME)
+            save_if_not_none(self._goal_state.shared_conf, SHARED_CONF_FILE_NAME)
+            save_if_not_none(self._goal_state.ext_conf, EXT_CONF_FILE_NAME.format(self._goal_state.incarnation))
+            save_if_not_none(self._goal_state.remote_access, REMOTE_ACCESS_FILE_NAME.format(self._goal_state.incarnation))
```

Comment:
Just curious, why is "Certificates.xml" handled differently in the goal_state.py file whereas all these other files are handled here?

Reply:
Yes, I forgot to add a comment here.

Certificates doesn't hold a reference to xml_text. I did not add one since I'm guessing the original intention was not to hold the data on memory for too long.